Signal/src/ViewControllers/PhotoLibrary/PhotoLibrary.swift

```diff
@@ -0,0 +1,288 @@
+
+                guard let exportSession = exportSession else {
+                    resolver.reject(PhotoLibraryError.assertionError(description: "exportSession was unexpectedly nil"))
+                    return
+                }
+
+                exportSession.outputFileType = AVFileType.mp4
+                exportSession.metadataItemFilter = AVMetadataItemFilter.forSharing()
+
+                let exportPath = OWSFileSystem.temporaryFilePath(withFileExtension: "mp4")
+                let exportURL = URL(fileURLWithPath: exportPath)
+                exportSession.outputURL = exportURL
+
+                Logger.debug("starting video export")
+                exportSession.exportAsynchronously {
+                    Logger.debug("Completed video export")
+
+                    guard let dataSource = DataSourcePath.dataSource(with: exportURL, shouldDeleteOnDeallocation: true) else {
+                        resolver.reject(PhotoLibraryError.assertionError(description: "Failed to build data source for exported video URL"))
+                        return
+                    }
+
+                    resolver.fulfill((dataSource: dataSource, dataUTI: kUTTypeMPEG4 as String))
+                }
+            }
+        }
+    }
+
+    func outgoingAttachment(for asset: PHAsset) -> Promise<SignalAttachment> {
+        switch asset.mediaType {
+        case .image:
+            return requestImageDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI, imageQuality: .medium)
+            }
+        case .video:
+            return requestVideoDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI)
+            }
+        default:
+            return Promise(error: PhotoLibraryError.unsupportedMediaType)
+        }
+    }
+}
+
+class PhotoCollection {
+    private let collection: PHAssetCollection
+
+    init(collection: PHAssetCollection) {
+        self.collection = collection
+    }
+
+    func localizedTitle() -> String {
+        guard let localizedTitle = collection.localizedTitle?.stripped,
+            localizedTitle.count > 0 else {
+            return NSLocalizedString("PHOTO_PICKER_UNNAMED_COLLECTION", comment: "label for system photo collections which have no name.")
+        }
+        return localizedTitle
+    }
+
+    func contents() -> PhotoCollectionContents {
+        let options = PHFetchOptions()
+        options.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: true)]
+        let fetchResult = PHAsset.fetchAssets(in: collection, options: options)
+
+        return PhotoCollectionContents(fetchResult: fetchResult, localizedTitle: localizedTitle())
+    }
+}
+
+class PhotoCollections {
```

Comment:
What do we gain from this class vs. passing around a `[PhotoCollection]`

Reply:
Done.